File: Commons/Graph/src/main/java/com/greentree/commons/graph/algorithm/walk/BFSWalker.java
package com.greentree.commons.graph.algorithm.walk;

import java.util.LinkedList;
import java.util.Queue;
import java.util.Stack;

import com.greentree.commons.graph.Graph;

public final class BFSWalker<V> implements GraphWalker<V> {
	
	
	private static final class BFSWalkInfo<V> {
		
		private final Stack<V> path = new Stack<>();
		
		private BFSWalkInfo() {
		}
		
		private BFSWalkInfo(BFSWalkInfo<V> p, V v) {
			path.addAll(p.path);
			path.add(v);
		}
		
		private BFSWalkInfo(V v) {
			path.add(v);
		}
		
		public boolean contains(V to) {
			return path.contains(to);
		}
		
		public V vertex() {
			return path.peek();
		}
		
	}
	
	private final Graph<? extends V> graph;
	
	public BFSWalker(Graph<? extends V> graph) {
		this.graph = graph;
	}
	
	@Override
	public void visit(V start, GraphVisitor<? super V> visitor) {
		final Queue<BFSWalkInfo<V>> queue = new LinkedList<>();
		visit(queue, visitor, new BFSWalkInfo<>(start));
		while(!queue.isEmpty()) {
			final var v = queue.remove();
			visit(queue, visitor, v);
		}
	}
	
	@Override
	public Graph<? extends V> graph() {
		return graph;
	}
	
	private void visit(Queue<BFSWalkInfo<V>> queue, GraphVisitor<? super V> visitor,
			BFSWalkInfo<V> info) {
		{
			final var v = info.vertex();
			var deadlock = true;
			for(var to : graph.getJoints(v))
				if(!info.contains(to)) {
					final var to_info = new BFSWalkInfo<>(info, to);
					queue.add(to_info);
					deadlock = false;
				}
			if(!deadlock)
				return;
		}
		GraphWalker.visitPath(info.path.iterator(), visitor);
	}
	
}

File: Commons/Graph/src/main/java/com/greentree/commons/graph/algorithm/walk/BFSWalker.kt
package com.greentree.commons.graph.algorithm.walk

import com.greentree.commons.graph.Graph
import com.greentree.commons.graph.algorithm.walk.path.PathVisitor
import com.greentree.commons.graph.algorithm.walk.vertex.VertexVisitor
import com.greentree.commons.graph.algorithm.walk.vertex.visitPath
import java.util.*

data class BFSWalker<V : Any>(private val graph: Graph<V>) : GraphWalker<V> {

	override fun visit(start: V, visitor: VertexVisitor<V>) {
		val queue: Queue<BFSVertexWalkInfo<V>> = LinkedList()
		visit(queue, visitor, BFSVertexWalkInfo(start))
		while(!queue.isEmpty()) {
			val v = queue.remove()
			visit(queue, visitor, v)
		}
	}

	override fun visit(start: V, visitor: PathVisitor<V>) {
		val queue: Queue<List<V>> = LinkedList()
		for(to in graph.getAdjacencySequence(start))
			queue.add(listOf(start, to))

		TODO("Not yet implemented")
	}

	private fun visit(
		queue: Queue<BFSVertexWalkInfo<V>>, visitor: VertexVisitor<V>,
		info: BFSVertexWalkInfo<V>,
	) {
		val v = info.vertex()
		var deadlock = true
		for(to in graph.getAdjacencyIterable(v))
			if(!info.contains(to)) {
				val toInfo = BFSVertexWalkInfo(info, to)
				queue.add(toInfo)
				deadlock = false
			}
		if(!deadlock)
			return
		visitor.visitPath(info.path.iterator())
	}

	private class BFSVertexWalkInfo<V> {

		val path: Stack<V> = Stack()

		private constructor()

		constructor(p: BFSVertexWalkInfo<V>, v: V) {
			path.addAll(p.path)
			path.add(v)
		}

		constructor(v: V) {
			path.add(v)
		}

		fun contains(to: V): Boolean {
			return path.contains(to)
		}

		fun vertex(): V {
			return path.peek()
		}
	}
}
